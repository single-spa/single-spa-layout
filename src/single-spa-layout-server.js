export { constructServerLayout } from "./server/constructServerLayout.js";
export { renderServerResponseBody } from "./server/renderServerResponseBody.js";
