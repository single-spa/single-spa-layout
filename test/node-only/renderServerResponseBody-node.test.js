import { constructServerLayout } from "../../src/single-spa-layout-server.js";
import { renderServerResponseBody } from "../../src/single-spa-layout-server.js";
import fs from "fs";
import path from "path";
import stream from "stream";

describe(`renderServerResponseBody`, () => {
  describe(`dom-elements.html fixture`, () => {
    it(`renders the app1 route correctly`, (done) => {
      const html = fs
        .readFileSync(
          path.resolve(process.cwd(), "./test/fixtures/dom-elements.html"),
          "utf-8"
        )
        .toString();

      const layout = constructServerLayout({
        html,
      });

      const readable = renderServerResponseBody(layout, {
        urlPath: "/app1",
        renderApplication(props) {
          const { name } = props;
          const appStream = new stream.Readable({
            read() {
              appStream.push(`<div id="single-spa-application:${name}"></div>`);
              appStream.push(null);
            },
          });
          return appStream;
        },
      });

      let finalHtml = "";

      readable.on("data", (chunk) => {
        finalHtml += chunk;
      });

      readable.on("end", () => {
        expect(finalHtml).toMatchSnapshot();
        done();
      });

      readable.read();
    });
  });

  it(`renders the app1 route fast`, (done) => {
    const html = fs
      .readFileSync(
        path.resolve(process.cwd(), "./test/fixtures/dom-elements.html"),
        "utf-8"
      )
      .toString();

    const layout = constructServerLayout({
      html,
    });

    const start = Date.now();
    const perfThresholdMs = 20;

    const readable = renderServerResponseBody(layout, {
      urlPath: "/app1",
      renderApplication(props) {
        const { name } = props;
        const appStream = new stream.Readable({
          read() {
            appStream.push(`<div id="single-spa-application:${name}"></div>`);
            appStream.push(null);
          },
        });
        return appStream;
      },
    });

    let finalHtml = "";

    readable.on("data", (chunk) => {
      finalHtml += chunk;
    });

    readable.on("end", () => {
      const latency = Date.now() - start;
      if (latency > perfThresholdMs) {
        fail(`It took ${latency}ms to render the server layout`);
      }
      done();
    });

    readable.read();
  });

  describe(`fragments.html fixture`, () => {
    it(`renders fragments correctly`, (done) => {
      const html = fs
        .readFileSync(
          path.resolve(process.cwd(), "./test/fixtures/fragments.html"),
          "utf-8"
        )
        .toString();

      const layout = constructServerLayout({
        html,
      });

      const readable = renderServerResponseBody(layout, {
        urlPath: "/app1",
        renderFragment(name) {
          const fragStream = new stream.Readable({
            read() {
              this.push(
                `
                <script type="systemjs-importmap">
                  {
                    "imports": {}
                  }
                </script>
              `.trim()
              );
              this.push(null);
            },
          });
          return fragStream;
        },
      });

      let finalHtml = "";

      readable.on("data", (chunk) => {
        finalHtml += chunk;
      });

      readable.on("end", () => {
        expect(finalHtml).toMatchSnapshot();
        done();
      });

      readable.read();
    });
  });
});
