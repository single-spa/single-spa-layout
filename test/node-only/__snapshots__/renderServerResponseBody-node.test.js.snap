// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderServerResponseBody dom-elements.html fixture renders the app1 route correctly 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Document
    </title>
  </head>
  <body>
    <div id="single-spa-application:header">
    </div>
    <div class="main-content">
      <button>
        Top level content
      </button>
      <div id="single-spa-application:app1">
      </div>
      Some raw text
    </div>
    <span>
      <footer>
        <a href="/about">
          About
        </a>
      </footer>
    </span>
    <single-spa-router mode="history"
                       base="/"
    >
      <div id="single-spa-application:undefined">
      </div>
      <div class="main-content">
        <button>
          Top level content
        </button>
        <div id="single-spa-application:undefined">
        </div>
        Some raw text
      </div>
      <span>
        <footer>
          <a href="/about">
            About
          </a>
        </footer>
      </span>
    </single-spa-router>
  </body>
</html>
`;

exports[`renderServerResponseBody fragments.html fixture renders fragments correctly 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, initial-scale=1.0"
    >
    <title>
      Document
    </title>
  </head>
  <body>
    <single-spa-router mode="history"
                       base="/"
    >
    </single-spa-router>
    <script type="systemjs-importmap">
      {
                    "imports": {}
                  }
    </script>
  </body>
</html>
`;
